crates/compiler/types/src/types.rs

```diff
@@ -589,33 +610,36 @@ impl Types {
     /// # Safety
     ///
     /// May only be called if `var` is known to represent the type at `index`.
-    pub unsafe fn set_variable(&mut self, index: Index<TypeTag>, var: Variable) {
-        self.tags[index.index()] = TypeTag::Variable(var);
+    pub unsafe fn emplace_variable(&self, index: Index<TypeCell>, var: Variable) {
+        self.tags[index.index()].0.replace(TypeTag::Variable(var));
     }
 
-    fn reserve_type_tags(&mut self, length: usize) -> Slice<TypeTag> {
+    fn reserve_type_tags(&mut self, length: usize) -> Slice<TypeCell> {
         use std::iter::repeat;
 
         debug_assert_eq!(self.tags.len(), self.tags_slices.len());
 
         self.tags_slices
             .extend(repeat(Slice::default()).take(length));
 
-        Slice::extend_new(&mut self.tags, repeat(TypeTag::EmptyRecord).take(length))
+        Slice::extend_new(
+            &mut self.tags,
+            repeat(TypeCell::new(TypeTag::EmptyRecord)).take(length),
+        )
     }
 
-    fn reserve_type_tag(&mut self) -> Index<TypeTag> {
+    fn reserve_type_tag(&mut self) -> Index<TypeCell> {
         debug_assert_eq!(self.tags.len(), self.tags_slices.len());
 
         self.tags_slices.push(Slice::default());
 
-        Index::push_new(&mut self.tags, TypeTag::EmptyRecord)
+        Index::push_new(&mut self.tags, TypeCell::new(TypeTag::EmptyRecord))
     }
 
-    fn set_type_tag(&mut self, index: Index<TypeTag>, tag: TypeTag, type_slice: Slice<TypeTag>) {
+    fn set_type_tag(&mut self, index: Index<TypeCell>, tag: TypeTag, type_slice: Slice<TypeCell>) {
         debug_assert_eq!(self.tags.len(), self.tags_slices.len());
 
-        self.tags[index.index()] = tag;
+        self.tags[index.index()].set(tag);
```

Comment:
you already have a mutable reference to `self.tags`, so why is the `Cell` required?